src/serializer/ResidualHeapVisitor.js

```diff
@@ -464,14 +469,39 @@ export class ResidualHeapVisitor {
     return binding;
   }
 
+  createGeneratorVisitCallbacks(generator: Generator): VisitEntryCallbacks {
+    return {
+      visitValue: this.visitValue.bind(this),
+      visitGenerator: this.visitGenerator.bind(this),
+      canSkip: (value: AbstractValue): boolean => {
+        return !this.referencedDeclaredValues.has(value) && !this.values.has(value);
+      },
+      recordDeclaration: (value: AbstractValue) => {
+        this.referencedDeclaredValues.add(value);
+      },
+      recordDelayedEntry: (entry: GeneratorEntry) => {
+        this.delayedVisitGeneratorEntries.push({ generator, entry });
+      },
+    };
+  }
+
   visitGenerator(generator: Generator): void {
     this._withScope(generator, () => {
-      generator.visit(this.visitValue.bind(this), this.visitGenerator.bind(this));
+      generator.visit(this.createGeneratorVisitCallbacks(generator));
     });
   }
 
   visitRoots(): void {
     this.visitGenerator(this.realmGenerator);
     for (let [, moduleValue] of this.modules.initializedModules) this.visitValue(moduleValue);
+    // Do a fixpoint over all pure generator entries to make sure that we visit
+    // arguments of only BodyEntries that are required by some other residual value
+    let oldDelayedEntries = [];
+    while (oldDelayedEntries.length !== this.delayedVisitGeneratorEntries.length) {
+      oldDelayedEntries = this.delayedVisitGeneratorEntries;
+      this.delayedVisitGeneratorEntries = [];
+      for (let { generator, entry } of oldDelayedEntries)
+        generator.visitEntry(entry, this.createGeneratorVisitCallbacks(generator));
```

Comment:
One more issue: This `visitEntry` call needs to happen within a `this._withScope(generator, () => {` block.